Make the scripts work regardless of the python version

MayeulC opened the pull request:
This builds upon, thus closes #9, which I rebased on upstream, and obsoletes #241. It also resolves #15.

This PR should work in theory, but please not that **I did not test it**. It would be great to setup CI with this repository.

I should also mention that I submitted similar patches to upstream ffmpeg. Unfortunately I cannot test the patches right now. Hopefully someone will chime in and do just that.

swick committed:
move proton from python2.7 to python3

MayeulC committed:
Make proton, gen_wrapper and user_settings python scripts version-agnostic

MayeulC commented on proton:
```diff
@@ -469,11 +470,11 @@ elif sys.argv[1] == "waitforexitandrun":
 elif sys.argv[1] == "getcompatpath":
     #linux -> windows path
     path = subprocess.check_output([wine_path, "winepath", "-w", sys.argv[2]], env=env, stderr=lfile)
-    sys.stdout.write(path.decode('utf-8'))
```
Please note that this was done to maximize compatibility with as few changes as necessary. @nanonyme expressed some objections to this change in https://github.com/ValveSoftware/Proton/pull/9#discussion_r211950495

The same change also happens four lines below.

MayeulC commented on proton:
```diff
@@ -469,11 +470,11 @@ elif sys.argv[1] == "waitforexitandrun":
 elif sys.argv[1] == "getcompatpath":
     #linux -> windows path
     path = subprocess.check_output([wine_path, "winepath", "-w", sys.argv[2]], env=env, stderr=lfile)
-    sys.stdout.write(path.decode('utf-8'))
```
Please note that this was done to maximize compatibility with as few changes as necessary. @nanonyme expressed some objections to this change in https://github.com/ValveSoftware/Proton/pull/9#discussion_r211950495

The same change also happens four lines below.

nanonyme replied:
Yes, it will only work in Python 2. Python 3 subprocess outputs bytes and stdin takes unicode. It will raise an exception as implicit decoding is no longer allowed

MayeulC replied:
mmm.. reverting this change actually doesn't seem to have any adverse effects

MayeulC commented on proton:
```diff
@@ -457,6 +458,13 @@ def run():
         else:
             run_wine([wine_path] + sys.argv[2:])
 
+if sys.version_info[0] == 2:
```
@nanonyme I finally bit the bullet and wrote this to mux the binary input to the correct binary output, according to your suggestion in https://github.com/ValveSoftware/Proton/pull/9#discussion_r212055974

I would have preferred to avoid explicitly different codepaths if it had peen possible.

Technically, it is possible to leave it as `sys.stdout.write(path.encode('utf-8'))`, but would cause weird issues on non utf8-encoded outputs that subprocess might spit out (or just outright fail).

nanonyme replied:
Also decoding is wrong since Python 2 expects bytes and decode results in unicode so you get implit encode inside write

MayeulC replied:
I meant `decode`, sorry. `encode` wouldn't work. But yeah, this would rely on implicit behaviour. Are you good with the changes now?

nanonyme replied:
This looks fine to me

nanonyme replied:
Note you can also
from __future__ import print_function

MayeulC replied:
Mmm, good point, that could remind people to keep the code compatible... On the other hand, there are very few `print` statements. I'd rather see a CI system that checks both with python 2 and 3.

nanonyme replied:
It's just that parenthesis does different things with print statement and function which can be confusing

aeikum commented:
@MayeulC @nanonyme Please let me know when you think this is ready to be merged and I'll review and test it. Thanks.

nanonyme commented:
I was just nitpicking with the print_function part, I'm pretty much done with comments.

MayeulC commented:
@nanonyme, I agreed with your last point, and I am going to add a commit to address this in a few minutes. @aeikum, it should be ready after that.

Edit: I gave credit to @nanonyme in the commit message. Please tell me if that's not OK.
I also didn't bother including this line in the `user_settings.sample.py` as I highly doubt this would be useful.

MayeulC committed:
Import print_function from __future__ for python compatibility

This alters the behaviour of python2's print statement to be compatible
with python3's print function. While not strictly necessary for now,
this might help future compatibility.
Thanks to Seppo Yli-Olli (@nanonyme) for suggesting this.

aeikum commented:
Thanks, this worked in both python2 and 3 for me.  This has been merged.
